File: sdks/typescript/src/integrations/langchain.ts
/**
 * LangChain.js integration for Layer5.
 *
 * LangChain.js uses handleToolStart / handleToolEnd / handleToolError
 * (not on_tool_start like Python).
 *
 * @example
 * ```ts
 * import { Layer5Callback } from '@layer5/sdk/integrations/langchain';
 * import { AgentExecutor } from 'langchain/agents';
 *
 * const callback = new Layer5Callback({
 *   apiKey:  'layer5_...',
 *   agentId: 'my-agent'
 * });
 *
 * const result = await executor.invoke(
 *   { input: 'fix the payment error' },
 *   { callbacks: [callback] }
 * );
 * ```
 */

import { Layer5 } from '../client.js';

interface ActiveCall {
  toolName: string;
  start: Date;
  context: Record<string, unknown>;
  decisionId?: string | null;
}

export class Layer5Callback {
  readonly name = 'Layer5Callback';

  private client: Layer5;
  private agentId: string;
  private silentErrors: boolean;
  private contextExtractor?: (
    toolName: string,
    toolInput: Record<string, unknown>
  ) => Record<string, unknown>;
  private activeCalls = new Map<string, ActiveCall>();

  constructor(options: {
    apiKey: string;
    agentId: string;
    baseUrl?: string;
    contextExtractor?: (
      toolName: string,
      toolInput: Record<string, unknown>
    ) => Record<string, unknown>;
    silentErrors?: boolean;
    timeout?: number;
    maxRetries?: number;
  }) {
    this.client = new Layer5({
      apiKey: options.apiKey,
      baseUrl: options.baseUrl,
      agentId: options.agentId,
      timeout: options.timeout,
      maxRetries: options.maxRetries,
    });
    this.agentId = options.agentId;
    this.silentErrors = options.silentErrors ?? true;
    this.contextExtractor = options.contextExtractor;
  }

  async handleToolStart(
    tool: { name?: string; [key: string]: unknown },
    input: string,
    runId: string,
    _parentRunId?: string,
    _tags?: string[],
    _metadata?: Record<string, unknown>,
    _name?: string,
  ): Promise<void> {
    const toolName = (tool.name ?? _name ?? 'unknown_tool') as string;

    let toolInput: Record<string, unknown> = {};
    try {
      toolInput = typeof input === 'string'
        ? JSON.parse(input)
        : (input as unknown as Record<string, unknown>);
    } catch {
      toolInput = { raw_input: String(input).slice(0, 500) };
    }

    const context = this.contextExtractor
      ? this.contextExtractor(toolName, toolInput)
      : { tool: toolName, ...toolInput };

    let decisionId: string | null | undefined;
    try {
      const scores = await this.client.getScores({
        agentId: this.agentId,
        context,
      });
      decisionId = scores?.decisionId;
    } catch (e) {
      this.handleError(e as Error, 'getScores failed');
    }

    this.activeCalls.set(runId, {
      toolName,
      start: new Date(),
      context,
      decisionId,
    });
  }

  async handleToolEnd(
    _output: string,
    runId: string,
  ): Promise<void> {
    const call = this.activeCalls.get(runId);
    this.activeCalls.delete(runId);
    if (!call) return;

    const responseMs = Date.now() - call.start.getTime();

    try {
      await this.client.logOutcome({
        agentId: this.agentId,
        actionName: call.toolName,
        success: true,
        context: call.context,
        responseMs,
        feedbackSignal: 'immediate',
        decisionId: call.decisionId ?? undefined,
      });
    } catch (e) {
      this.handleError(e as Error, 'logOutcome failed');
    }
  }

  async handleToolError(
    _error: Error,
    runId: string,
  ): Promise<void> {
    const call = this.activeCalls.get(runId);
    this.activeCalls.delete(runId);
    if (!call) return;

    const responseMs = Date.now() - call.start.getTime();

    try {
      await this.client.logOutcome({
        agentId: this.agentId,
        actionName: call.toolName,
        success: false,
        context: call.context,
        responseMs,
        feedbackSignal: 'immediate',
        decisionId: call.decisionId ?? undefined,
      });
    } catch (e) {
      this.handleError(e as Error, 'logOutcome(error) failed');
    }
  }

  private handleError(error: Error, context: string): void {
    if (!this.silentErrors) throw error;
    console.warn(
      `[Layer5] ${context}: ${error.message}. ` +
      `Agent continues normally.`
    );
  }
}

File: sdks/python/layer5/integrations/langchain.py
"""
LangChain integration for Layer5.

Usage:
    from langchain.agents import AgentExecutor
    from layer5.integrations.langchain import Layer5Callback

    callback = Layer5Callback(
        api_key="layer5_...",
        agent_id="my-langchain-agent"
    )

    agent = AgentExecutor(
        agent=agent,
        tools=tools,
        callbacks=[callback]
    )

    # That's it. Layer5 now scores every tool call.
"""

from __future__ import annotations

import json
import warnings
from datetime import datetime
from typing import Any, Callable, Dict, List, Optional, Union

from ..client import Layer5
from ..exceptions import Layer5Error

try:
    from langchain_core.callbacks import BaseCallbackHandler

    LANGCHAIN_AVAILABLE = True
except ImportError:
    try:
        from langchain.callbacks.base import BaseCallbackHandler  # type: ignore[no-redef]

        LANGCHAIN_AVAILABLE = True
    except ImportError:
        LANGCHAIN_AVAILABLE = False
        BaseCallbackHandler = object  # type: ignore[assignment,misc]


class Layer5Callback(BaseCallbackHandler if LANGCHAIN_AVAILABLE else object):  # type: ignore[misc]
    """
    LangChain callback handler for Layer5.
    Automatically intercepts every tool call:
      - Fetches scores before tool execution
      - Logs outcome after tool execution
      - Handles errors gracefully (never crashes your agent)

    Works with:
      - AgentExecutor
      - LangGraph
      - Any LangChain tool-using pattern
    """

    def __init__(
        self,
        api_key: str,
        agent_id: str,
        base_url: str = "https://api.layer5.dev",
        context_extractor: Optional[Callable[[dict], dict]] = None,
        silent_errors: bool = True,
    ):
        """
        Args:
            api_key:           Your Layer5 API key
            agent_id:          Identifier for this agent
            context_extractor: Optional function to extract context from
                              tool input dict.
                              Signature: (tool_input: dict) -> dict
                              Default: passes tool_input as context
            silent_errors:     If True (default), Layer5 errors never crash
                              your agent. They are logged as warnings
                              instead. Set False in testing to surface
                              errors.
        """
        if not LANGCHAIN_AVAILABLE:
            raise ImportError(
                "LangChain is required for Layer5Callback. "
                "Install it: pip install langchain-core"
            )

        self.client = Layer5(
            api_key=api_key,
            base_url=base_url,
            agent_id=agent_id,
        )
        self.agent_id = agent_id
        self.context_extractor = context_extractor
        self.silent_errors = silent_errors

        # Track in-flight tool calls
        # run_id → {tool_name, start_time, context, scores}
        self._active_calls: Dict[str, dict] = {}

    def on_tool_start(
        self,
        serialized: Dict[str, Any],
        input_str: str,
        *,
        run_id: Any,
        **kwargs: Any,
    ) -> None:
        """Called before every tool execution."""
        tool_name = serialized.get("name", "unknown_tool")

        # Extract context from tool input
        try:
            tool_input = (
                json.loads(input_str) if isinstance(input_str, str) else {}
            )
        except Exception:
            tool_input = {"raw_input": str(input_str)[:500]}

        context = (
            self.context_extractor(tool_input)
            if self.context_extractor
            else tool_input
        )

        # Ensure issue_type is present for the API
        if "issue_type" not in context:
            context["issue_type"] = tool_name

        # Fetch scores (non-blocking on error)
        scores = None
        try:
            scores = self.client.get_scores(
                agent_id=self.agent_id,
                context=context,
            )
        except Layer5Error as e:
            if not self.silent_errors:
                raise
            warnings.warn(
                f"Layer5 get_scores failed (agent will continue): {e}",
                stacklevel=2,
            )

        self._active_calls[str(run_id)] = {
            "tool_name": tool_name,
            "start_time": datetime.now(),
            "context": context,
            "scores": scores,
            "decision_id": scores.decision_id if scores else None,
        }

    def on_tool_end(
        self,
        output: str,
        *,
        run_id: Any,
        **kwargs: Any,
    ) -> None:
        """Called after successful tool execution."""
        call = self._active_calls.pop(str(run_id), None)
        if not call:
            return

        elapsed_ms = int(
            (datetime.now() - call["start_time"]).total_seconds() * 1000
        )

        try:
            self.client.log_outcome(
                agent_id=self.agent_id,
                action_name=call["tool_name"],
                session_id="langchain-auto",
                issue_type=call["context"].get("issue_type", call["tool_name"]),
                success=True,
                raw_context=call["context"],
                response_time_ms=elapsed_ms,
                feedback_signal="immediate",
                decision_id=call.get("decision_id"),
            )
        except Layer5Error as e:
            if not self.silent_errors:
                raise
            warnings.warn(
                f"Layer5 log_outcome failed: {e}",
                stacklevel=2,
            )

    def on_tool_error(
        self,
        error: Union[Exception, KeyboardInterrupt],
        *,
        run_id: Any,
        **kwargs: Any,
    ) -> None:
        """Called when a tool raises an exception."""
        call = self._active_calls.pop(str(run_id), None)
        if not call:
            return

        elapsed_ms = int(
            (datetime.now() - call["start_time"]).total_seconds() * 1000
        )

        try:
            self.client.log_outcome(
                agent_id=self.agent_id,
                action_name=call["tool_name"],
                session_id="langchain-auto",
                issue_type=call["context"].get("issue_type", call["tool_name"]),
                success=False,
                raw_context=call["context"],
                response_time_ms=elapsed_ms,
                feedback_signal="immediate",
                decision_id=call.get("decision_id"),
            )
        except Layer5Error as e:
            if not self.silent_errors:
                raise
            warnings.warn(
                f"Layer5 log_outcome (error) failed: {e}",
                stacklevel=2,
            )
